### PR TITLE
Improve security & performance of Encrypt then MAC scheme.

### PR DESCRIPTION
**Pull request by lynn-stephenson**
First off, this will properly split a single symmetric key into multiple, to increase the complexity of the relationship between the encryption & MAC key.

I removed unnecessary code (resulting in less dependencies, increasing simplicity), and documentation.

This Pull request is not ready to be merged, due to the fact I have not implemented unit tests nor I have I signed my work. I'm requesting review of this draft, prior to submitting the PR that should be accepted.

With that said, how should we go about the unit tests?

### CODE REVIEW

File: lib/private/Security/Crypto.php [hunk 1]
```diff
@@ -90,16 +82,19 @@ public function encrypt(string $plaintext, string $password = ''): string {
 		if ($password === '') {
 			$password = $this->config->getSystemValue('secret');
 		}
-		$this->cipher->setPassword($password);
+		
+		$keyMaterial = hash_hkdf('sha512', $password);
```
Comment (kesselb) [inline lib/private/Security/Crypto.php:86]:
```suggestion
		$keyMaterial = hash_hkdf('sha512', $password, 32);
```

I'm not 100% sure. The [documentation](https://www.php.net/manual/en/function.hash-hkdf.php
) says "Desired output length in bytes" :man_shrugging:

**Review by kesselb — COMMENTED** [review]
Thanks for your pull request :+1:

**Comment by kesselb** [comment]
> With that said, how should we go about the unit tests?

The failing tests at https://drone.nextcloud.com/nextcloud/server/34111 could be a start.

File: lib/private/Security/Crypto.php (hunk 1 above)
Comment (lynn-stephenson) [reply lib/private/Security/Crypto.php:86]: @kesselb The documentation says `hash_hkdf` with a `$length` of `0` will use the provided hash function length. Also, 32 bytes is not enough. If anything, we'll explicitly ask for 64 bytes.

**Comment by lynn-stephenson** [comment]
It will take me a bit to set up my dev env, because my OS doesn't support stuff NextCloud needs.

But ill dig into the unit tests as soon as i can. Obviously gonna have to modify the Crypto tests, but there appears to be issues with the Storage tests, which may be checking for ciphertext outputs too.

**Comment by kesselb** [comment]
> It will take me a bit to set up my dev env, because my OS doesn't support stuff NextCloud needs.

https://github.com/juliushaertl/nextcloud-docker-dev/ might help with the dev setup

**Comment by lynn-stephenson** [comment]
@kesselb I managed to get my env set up, it would was good I did it anyways. i need to learn docker at some point.

Ill try to figure out these unit tests fails tomorrow if i got time.

**Comment by MorrisJobke** [comment]
> Ill try to figure out these unit tests fails tomorrow if i got time.

We are currently looking into some failures that are unrelated. So if they are not obviously related in the first place, then maybe they are just false positives (see also #22305)

**Comment by MorrisJobke** [comment]
But those seem related :D 

https://drone.nextcloud.com/nextcloud/server/34111/8/3

**Pull request closed without merging**